modules/system/classes/extensions/source/ExtensionSource.php

```diff
@@ -0,0 +1,354 @@
+use System\Classes\Extensions\ExtensionManagerInterface;
+use System\Classes\Extensions\ModuleManager;
+use System\Classes\Extensions\PluginManager;
+use Winter\Storm\Foundation\Extension\WinterExtension;
+use Winter\Storm\Packager\Composer;
+use Winter\Packager\Exceptions\CommandException;
+use Winter\Storm\Exception\ApplicationException;
+use Winter\Storm\Support\Str;
+
+class ExtensionSource
+{
+    public const SOURCE_COMPOSER = 'composer';
+    public const SOURCE_MARKET = 'market';
+    public const SOURCE_LOCAL = 'local';
+
+    public const TYPE_PLUGIN = 'plugin';
+    public const TYPE_THEME = 'theme';
+    public const TYPE_MODULE = 'module';
+
+    public const STATUS_UNINSTALLED = 'uninstalled';
+    public const STATUS_INSTALLED = 'installed';
+    public const STATUS_UNPACKED = 'unpacked';
+
+    protected array $extensionManagerMapping = [
+        self::TYPE_PLUGIN => PluginManager::class,
+        self::TYPE_THEME => ThemeManager::class,
+        self::TYPE_MODULE => ModuleManager::class,
+    ];
+
+    protected string $status = 'uninstalled';
+
+    public function __construct(
+        public string $source,
+        public string $type,
+        public ?string $code = null,
+        public ?string $composerPackage = null,
+        public ?string $path = null
+    ) {
+        if (!in_array($this->source, [static::SOURCE_COMPOSER, static::SOURCE_MARKET, static::SOURCE_LOCAL])) {
+            throw new \InvalidArgumentException("Invalid source '{$this->source}'");
+        }
+
+        if (!in_array($this->type, [static::TYPE_PLUGIN, static::TYPE_THEME, static::TYPE_MODULE])) {
+            throw new \InvalidArgumentException("Invalid type '{$this->type}'");
+        }
+
+        if ($this->source === static::SOURCE_COMPOSER && !$this->composerPackage) {
+            throw new ApplicationException('You must provide a composer package for a composer source.');
+        }
+
+        if ($this->source !== static::SOURCE_COMPOSER && !$this->code) {
+            if (!$this->path) {
+                throw new ApplicationException('You must provide a code or path.');
+            }
+
+            $this->code = $this->guessCodeFromPath($this->path);
+        }
+
+        $this->status = $this->checkStatus();
+    }
+
+    public function getStatus(): string
+    {
+        return $this->status;
+    }
+
+    public function getCode(): ?string
+    {
+        if ($this->code) {
+            return $this->code;
+        }
+
+        if (!$this->path) {
+            return null;
+        }
+
+        return $this->code = $this->guessCodeFromPath($this->path);
+    }
+
+    public function getPath(): ?string
+    {
+        if ($this->path) {
+            return $this->path;
+        }
+
+        if (!$this->code) {
+            return null;
+        }
+
+        return $this->path = $this->guessPathFromCode($this->code);
+    }
+
+    /**
+     * @throws ApplicationException
+     */
+    public function createFiles(): ?static
```

Comment:
Why not move this logic to its dedicated ExtensionSource classes (i.e. ComposerSource, LocalSource, MaketSource)

Reply:
Honestly because I only added those classes later after already implementing all of this logic in `ExtensionSource`, but yes I think it should be moved. Was just easier during development to keep them all in the same place.

Reply:
Ah, so we might not want to do this as the internal `$source` property is changed during install, i.e. after `composer require ...` is executed and we have the files, we modify the `$source` to local so we don't re-download if `createFiles()` is called again. Just did the work to split it all out and not sure if I like it :thinking:

Reply:
Ah, makes sense to keep it in here then!